Add Volume Health Support to CSI (#415)
container-storage-interface/spec: spec.md

  rpc ControllerExpandVolume (ControllerExpandVolumeRequest)
    returns (ControllerExpandVolumeResponse) {}

>   rpc GetVolume (GetVolumeRequest)

Reviewer: `GetVolume` -> `ControllerGetVolume`

|-----------|-----------|-------------|-------------------|
| Invalid `starting_token` | 10 ABORTED | Indicates that `starting_token` is not valid. | Caller SHOULD start the `ListVolumes` operation again with an empty `starting_token`. |

#### `GetVolume`

A Controller Plugin MUST implement this `GetVolume` RPC call if it has `GET_VOLUME` capability.

> A Controller Plugin MUST support the `volume_health` field in the `status` field in `GetVolumeResponse` if it has `GET_VOLUME_HEALTH` capability.

Reviewer 2: I question the MUST wording here. The CO needs to be programmed defensively against UNKNOWN health conditions anyway, so why force a validation case here. IOW how will the CO reason differently about a missing health object (when VOLUME_HEALTH cap is reported) vs. a health object w/ UNKNOWN volume condition?

Instead, I'd like for us to consider defining `volume_health` as *always* optional, with the wording stating that plugins SHOULD provide a non-empty `volume_health` field if they advertise the `VOLUME_HEALTH` capability. This steers plugins toward the behavior that we want. The CO still has to handle both of the aforementioned cases, but neither case is now perceived an an "error" on the plugin's part.

Author: Sure

> `GetVolumeResponse` should contain current information of a volume if it exists. If the volume does not exist any more, `GetVolume` should return gRPC error code `NOT_FOUND`.

Reviewer 2: nit: one sentence per line, please

The `VolumeHealth` message in `VolumeStatus` field in `GetVolumeResponse` can indicate additional health information for the volume.

> Whether the volume is still attached to a node is already handled by `LIST_VOLUMES_PUBLISHED_NODES` controller capability. This capability MAY be used by CO to reconcile - re-attach the volume if the volume is not attached any more.

Reviewer 2: nit: one sentence per line please

Author: Sure

```protobuf
message GetVolumeRequest {
  // Identity information for a specific volume. This field is
  // REQUIRED. GetVolume will return with current volume information.
  string volume_id = 1;
}

message GetVolumeResponse {
  message VolumeStatus{
    // volume_health shows error conditions reported by the SP.
    // This field MUST be specified if the
    // GET_VOLUME_HEALTH controller capability is supported.
    repeated VolumeHealth volume_health = 1;
  }

  // This field is REQUIRED
  Volume volume = 1;

  // This field is OPTIONAL. This field MUST be specified if the
  // GET_VOLUME_HEALTH controller capability is supported.
  VolumeStatus status = 2;
}
```

##### GetVolume Errors

If the plugin is unable to complete the GetVolume call successfully, it MUST return a non-ok gRPC code in the gRPC status.

##### VolumeHealth from Controller vs Node Plugins

> If a SP provides `VolumeHealth` information in both Controller and Node Plugins, information provided SHALL not be duplicate. `VolumeHealth` information from the Controller Plugin SHALL be from the storage system but not specific for one node, while information from the Node Plugin SHALL be specific for that node.

Reviewer 2: "SHALL not be duplicate" -- what if it is? is the CO expected to fail the RPC call? this sounds arbitrarily restrictive, though I think I understand the goal of this wording. please re-phrase.

nit: one sentence per line please

Author: Sure

    // published_node_ids MAY include nodes not published to or
    // reported by the SP. The CO MUST be resilient to that.
    repeated string published_node_ids = 1;

>     // Information about the current health of the volume.

Reviewer: Per @jdef's suggestion https://github.com/container-storage-interface/spec/pull/415#discussion_r406571188:
```suggestion
    // Information about the current condition of the volume.
```

A Controller Plugin SHOULD provide a non-empty `volume_health` field in `GetVolumeResponse` if it has `VOLUME_HEALTH` capability.

`GetVolumeResponse` should contain current information of a volume if it exists.
If the volume does not exist any more, `GetVolume` should return gRPC error code `NOT_FOUND`.

The `VolumeHealth` message in `VolumeStatus` field in `GetVolumeResponse` can indicate additional health information for the volume.

Whether the volume is still attached to a node is already handled by `LIST_VOLUMES_PUBLISHED_NODES` controller capability.
This capability MAY be used by CO to reconcile - re-attach the volume if the volume is not attached any more.

```protobuf
message GetVolumeRequest {
  option (alpha_message) = true;

  // Identity information for a specific volume. This field is
  // REQUIRED. GetVolume will return with current volume information.
  string volume_id = 1;
}

message GetVolumeResponse {
  option (alpha_message) = true;

  message VolumeStatus{
    // volume_health shows error conditions reported by the SP.
    // This field MUST be specified if the
    // VOLUME_HEALTH controller capability is supported.

>     VolumeHealth volume_health = 1 [(alpha_field) = true];

Reviewer 2: this field appears within an embedded message type whose parent type is already annotated as alpha. this seems redundant and as such is not aligned with the proposal in #365 

read: please remove the alpha annotation here, unless there's strong argument in favor of it

Author: Removed

  }

  // This field is REQUIRED
  Volume volume = 1;

  // This field is OPTIONAL. This field MUST be specified if the
  // VOLUME_HEALTH controller capability is supported.
  VolumeStatus status = 2;
}
```

##### GetVolume Errors

If the plugin is unable to complete the GetVolume call successfully, it MUST return a non-ok gRPC code in the gRPC status.

##### VolumeHealth from Controller vs Node Plugins

A SP can implement `VolumeHealth` in only the Controller Plugin, only the Node Plugin, or both if possible.

> If `VolumeHealth` is implemented in both the Controller and Node Plugins, it SHALL try to avoid confusion by reporting from different perspectives. 

Reviewer 2: Is the intent that the node perspective always overrides that of the controller, or vice-versa?

Or, instead, should CO's be pessimistic and if controller/node are reporting different health conditions then the CO should assume the worst case is the truth?

Asking the CO to merge knowledge of varying health conditions for the same volume seems perilous, but if that's what is expected here then we should probably include a recommendation for COs as to how they should interpret misaligned health conditions (controller and node appear to disagree).

Author: Added comments to clarify this.  If controller plugin says the storage system is healthy but node plugin reports there is a mounting problem, they are from different perspective so both are correct.  If in any case they are misaligned, CO should alway pay attention to the worst case.

```protobuf
message ControllerGetVolumeRequest {
  option (alpha_message) = true;

  // The ID of the volume to fetch current volume information for.
  // This field is REQUIRED.
  string volume_id = 1;
}

message ControllerGetVolumeResponse {
  option (alpha_message) = true;

  message VolumeStatus{
    // A list of all the `node_id` of nodes that this volume is
    // controller published on.
    // This field is OPTIONAL.
    // This field MUST be specified if the PUBLISH_UNPUBLISH_VOLUME
    // controller capability is supported.
    // published_node_ids MAY include nodes not published to or
    // reported by the SP. The CO MUST be resilient to that.
    repeated string published_node_ids = 1;

    // Information about the current condition of the volume.
    // This field is OPTIONAL.
    // This field MUST be specified if the
    // VOLUME_CONDITION controller capability is supported.
    VolumeCondition volume_condition = 2;
  }

  // This field is REQUIRED
  Volume volume = 1;

  // This field is REQUIRED.
  VolumeStatus status = 2;
}
```

##### ControllerGetVolume Errors

> If the plugin is unable to complete the ControllerGetVolume call successfully, it MUST return a non-ok gRPC code in the gRPC status. If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code. The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.

Author: Done

A Controller Plugin SHOULD provide a non-empty `volume_health` field in `GetVolumeResponse` if it has `VOLUME_HEALTH` capability.

`GetVolumeResponse` should contain current information of a volume if it exists.
If the volume does not exist any more, `GetVolume` should return gRPC error code `NOT_FOUND`.

The `VolumeHealth` message in `VolumeStatus` field in `GetVolumeResponse` can indicate additional health information for the volume.

Whether the volume is still attached to a node is already handled by `LIST_VOLUMES_PUBLISHED_NODES` controller capability.
This capability MAY be used by CO to reconcile - re-attach the volume if the volume is not attached any more.

```protobuf
message GetVolumeRequest {
  option (alpha_message) = true;

  // Identity information for a specific volume. This field is
  // REQUIRED. GetVolume will return with current volume information.
  string volume_id = 1;
}

message GetVolumeResponse {
  option (alpha_message) = true;

  message VolumeStatus{
    // volume_health shows error conditions reported by the SP.
    // This field MUST be specified if the
    // VOLUME_HEALTH controller capability is supported.

>     VolumeHealth volume_health = 1;

Author: Sure

>     VolumeHealth volume_health = 1;

Author: Done

A Controller Plugin MUST provide a non-empty `volume_condition` field in `ControllerGetVolumeResponse` if it has `VOLUME_CONDITION` capability.

`ControllerGetVolumeResponse` should contain current information of a volume if it exists.
If the volume does not exist any more, `ControllerGetVolume` should return gRPC error code `NOT_FOUND`.

```protobuf
message ControllerGetVolumeRequest {
  option (alpha_message) = true;

  // Identity information for a specific volume. This field is
  // REQUIRED. ControllerGetVolume will return with current
  // volume information.
  string volume_id = 1;
}

message ControllerGetVolumeResponse {
  option (alpha_message) = true;

  message VolumeStatus{
    // A list of all the `node_id` of nodes that this volume is
    // controller published on.
    // This field is OPTIONAL.
    // This field MUST be specified if the PUBLISH_UNPUBLISH_VOLUME
    // controller capability is supported.
    // published_node_ids MAY include nodes not published to or
    // reported by the SP. The CO MUST be resilient to that.
    repeated string published_node_ids = 1;

>     // Information about the current health of the volume.

Reviewer: Per @jdef's suggestion https://github.com/container-storage-interface/spec/pull/415#discussion_r406571188:
```suggestion
    // Information about the current condition of the volume.
```

|-----------|-----------|-------------|-------------------|
| Invalid `starting_token` | 10 ABORTED | Indicates that `starting_token` is not valid. | Caller SHOULD start the `ListVolumes` operation again with an empty `starting_token`. |

#### `GetVolume`

**EXPERIMENTAL FEATURE**

A Controller Plugin MUST implement this `GetVolume` RPC call if it has `GET_VOLUME` capability.

> A Controller Plugin SHOULD provide a non-empty `volume_health` field in `GetVolumeResponse` if it has `VOLUME_HEALTH` capability.

Author: Done

  message VolumeStatus{
    // A list of all `node_id` of nodes that the volume in this entry
    // is controller published on.
    // This field is OPTIONAL. If it is not specified and the SP has
    // the PUBLISH_UNPUBLISH_VOLUME controller capability, the CO
    // MAY assume the volume is not controller published to any nodes.
    // If the field is not specified and the SP does not have the
    // PUBLISH_UNPUBLISH_VOLUME controller capability, the CO MUST
    // not interpret this field.
    // published_node_ids MAY include nodes not published to or
    // reported by the SP. The CO MUST be resilient to that.
    repeated string published_node_ids = 1;

    // volume_condition shows information about the current health of
    // the volume.
    // This field is OPTIONAL.
    // This field MUST be specified if the
    // VOLUME_CONDITION controller capability is supported.
    VolumeCondition volume_condition = 2;
  }

  // This field is REQUIRED
  Volume volume = 1;

  // This field is OPTIONAL. This field MUST be specified if the
  // VOLUME_CONDITION controller capability is supported.
  VolumeStatus status = 2;
}
```

##### ControllerGetVolume Errors

> If the plugin is unable to complete the ControllerGetVolume call successfully, it MUST return a non-ok gRPC code in the gRPC status.

Author: Done

##### VolumeHealth from Controller vs Node Plugins

A SP can implement `VolumeHealth` in only the Controller Plugin, only the Node Plugin, or both if possible.

If `VolumeHealth` is implemented in both the Controller and Node Plugins, it SHALL report from different perspectives.
For example, `VolumeHealth` information from the Controller Plugin SHALL be from the storage system but not specific for one node, while information from the Node Plugin SHALL be specific for that node.

If for some reason Controller and Node Plugins report misaligned health conditions, CO SHALL assume the worst case is the truth.

> Note that for this initial implementation of `VolumeHealth` support, the intended consumer of signal should be informative for humans, not for automated machine response.

Author: Sure

      // Indicates the SP supports the
      // ListVolumesResponse.entry.published_nodes field
      LIST_VOLUMES_PUBLISHED_NODES = 10;

      // Indicates the SP supports the GetVolume RPC
      GET_VOLUME = 11;

      // Indicates the SP supports the
      // GetVolumeResponse.volume_health field

>       GET_VOLUME_HEALTH = 12;

Author: Done

      // Indicates the SP supports the
      // ListVolumesResponse.entry.published_nodes field
      LIST_VOLUMES_PUBLISHED_NODES = 10;

      // Indicates the SP supports the
      // ListVolumesResponse.entry.volume_health field

>       LIST_VOLUMES_VOLUME_HEALTH = 11;

Reviewer 2: instead of adding new capabilities here, the Alpha API proposal suggests a separate AlphaFeature bucket. See https://github.com/container-storage-interface/spec/pull/365/commits/105831102506ec4999e8386c08a3421bed0b6d9f#diff-308dbd5d943fd4b49fc73f1092c253abR948

Author: Done

      // Indicates the SP supports the
      // ListVolumesResponse.entry.published_nodes field
      LIST_VOLUMES_PUBLISHED_NODES = 10;

>       // Indicates the SP supports controller level volume condition.

Author: Done

      // Indicates the SP supports the
      // ListVolumesResponse.entry.published_nodes field
      LIST_VOLUMES_PUBLISHED_NODES = 10;

      // Indicates the SP supports controller level volume condition.
      // An SP can implement `VolumeCondition` in only the Controller
      // Plugin, only the Node Plugin, or both if possible.
      // If `VolumeCondition` is implemented in both the Controller and
      // Node Plugins, it SHALL report from different perspectives.
      // If for some reason Controller and Node Plugins report
      // misaligned volume conditions, CO SHALL assume the worst case
      // is the truth.

>       // Note that for for alpha `VolumeCondition` is intended be
>       // informative for humans only, not for automated machine
>       // response.

Reviewer: ```suggestion
      // Note that, for alpha, `VolumeCondition` is intended be
      // informative for humans only, not for automation.
```

      VOLUME_CONDITION = 11 [(alpha_enum_value) = true];

      // Indicates the SP supports the ControllerGetVolume RPC.
      // This enables COs to, for example, fetch per volume

>       // health after a volume is provisioned.

Reviewer: Per @jdef's suggestion https://github.com/container-storage-interface/spec/pull/415#discussion_r406571188:
```suggestion
      // condition after a volume is provisioned.
```

      // Indicates the SP supports the
      // ListVolumesResponse.entry.published_nodes field
      LIST_VOLUMES_PUBLISHED_NODES = 10;

      // Indicates the SP supports the
      // ListVolumesResponse.entry.volume_condition field
      LIST_VOLUMES_VOLUME_CONDITION = 11 [(alpha_enum_value) = true];

      // Indicates the SP supports the GetVolume RPC
      GET_VOLUME = 12 [(alpha_enum_value) = true];

      // Indicates the SP supports the
      // GetVolumeResponse.volume_condition field

>       GET_VOLUME_VOLUME_CONDITION = 13 [(alpha_enum_value) = true];

Reviewer: Instead of both `LIST_VOLUMES_VOLUME_CONDITION` and `GET_VOLUME_VOLUME_CONDITION` let's just have one -- `VOLUME_CONDITION`. Caller can always check for `LIST_VOLUMES` and `GET_VOLUME` capability.

      // Indicates the SP supports the
      // ListVolumesResponse.entry.published_nodes field
      LIST_VOLUMES_PUBLISHED_NODES = 10;

      // Indicates the SP supports the
      // ListVolumesResponse.entry.status.volume_condition field
      // if SP also supports LIST_VOLUMES capability, or supports the
      // ControllerGetVolumeResponse.status.volume_condition field

>       // if SP also supports GET_VOLUME capability

Author: Done

      VOLUME_CONDITION = 11 [(alpha_enum_value) = true];

>       // Indicates the SP supports the ControllerGetVolume RPC

Author: Done

  // Units by which values are measured. This field is REQUIRED.
  Unit unit = 4;
}

message VolumeHealth {

>   enum Severity {

Author: Done

    // Volume health condition is unknown; treat it as healthy.
    UNKNOWN = 0;
    // Volume is healthy.
    HEALTHY = 1;
    // Volume is temporarily unhealthy, but will likely become healthy.
    TEMPORARY_UNHEALTHY = 2;
    // Volume will fail soon. Move your data off and replace volume or
    // you may lose your data.

>     FATAL_EMINENT = 3;

Author: Done

    // Volume is unusable. Data loss is likely.
    FATAL = 4;
  }

  // The severity level of the health condition of the volume.
  // This field is REQUIRED.
  Severity severity = 1;

  // The error code describing the health condition of the volume.
  // This is an opaque field to CO.
  // This field is REQUIRED.

>   string error_code = 2;

Author: My understanding from last meeting is that Severity/Condition is a field CO can understand, but error code is still important for user to see what kind of error it is and it is specific to the storage systems.

>   string error_code = 2;

Reviewer 3: How useful is error code if the SP can define their own?

>   string error_code = 2;

Author: Admin of the storage system should understand the error code and decides what to do with it so it is useful.

Plugins can use this field to determine if `volume_path` is where the volume is published or staged,
and setting this field to non-empty allows plugins to function with less stored state on the node.

> A Node plugin MUST implement NodeGetVolumeStats RPC call for fetching volume condition if it has VOLUME_CONDITION node capability.

Author: Done

> A Node plugin MUST implement NodeGetVolumeStats RPC call for fetching volume condition if it has VOLUME_CONDITION node capability.

Reviewer: Please remove now that you've combined with the first line.

  // Units by which values are measured. This field is REQUIRED.
  Unit unit = 4;
}

message VolumeHealth {
  option (alpha_message) = true;

>   enum Condition {

Reviewer 2: should we number these conditions with room to spare, in case we need to add intermediate values? at a glance these look like they are sorted by severity (minus UNKNOWN). if that's an idiom that we want to maintain for this enumeration, maybe number them like 10, 20, 30, 40 in case we think of a new value we want to squeeze in between?

or not .. this is really just food for thought, not a hard requirement IMO

Author: That's a good idea.  I'll use 10, 20, ...

>     // Volume health condition is unknown; If this is the initial

Reviewer 2: initial w/ respect to ... plugin startup? something else? i think i understand what's being attempted here - some sort of suggestion w/ respect to state machine transitions - and it should be probably made even more clear in such case.

Author: Maybe it's better just to say "Volume health condition is unknown" here and remove any other comments?

    // report of volume health, treat it as healthy. If there was
    // already a condition reported previously, treat unknown as
    // unchanged from the previous condition.
    UNKNOWN = 0;
    // Volume is accessible and no problems are detected. 
    HEALTHY = 1;
    // Volume is not accessible currently, but there's no reason to
    // expect data loss. Examples: power outage, network outage,
    // maintenance.
    TEMPORARY_INACCESSIBLE = 2;
    // Volume is accessible currently, but is degraded. Data loss

>     // is unlikely as the storage system is rebuilding to fix the

Reviewer 2: s/is rebuilding/may be rebuilding/

it actually may not be automatic, and operator intervention may be required here - right? if not, and the intent is to differentiate between states that automation will recover vs. those that a human will recover then we probably need another state here.

i'm leaning towards NOT adding a new state to differentiate, and simply update the wording here instead.

Author: Yes, it is possible that operator intervention may be needed.  I changed to "may be rebuilding".

    // problem automatically. For example, a replica is lost on
    // the storage system but a spare is configured, and the storage
    // system is resyncing/reconstructing.
    TEMPORARY_DEGRADED = 3;
    // Volume is accessible currently, but a problem is detected that
    // could lead to eventual data loss. Examples: recoverable hardware
    // failure (i.e. the disk needs to be manually replaced by a human),
    // cooling failure, equipment significantly past end of life.
    FAILURE_LIKELY = 4;
    // Volume is not accessible currently, and there is reason to
    // believe the outage may be permanent. Examples: unrecoverable
    // hardware failure, natural disasters, damage to equipment.
    FATAL = 5;
  }

  // The level of the health condition of the volume.
  // This field is REQUIRED.
  Condition condition = 1;

  // The error code and message describing the health condition of
  // the volume. This field is REQUIRED.
  repeated VolumeHealthError error = 2;
}

> message VolumeHealthError {

Reviewer 2: any reason this message type isn't embedded within VolumeHealth? if it is to stay separate, then it needs an "alpha" annotation

Author: I'll move it inside VolumeHealth.

  // The error code describing the health condition of the volume.
  // This is an opaque field to CO.
  // This field is REQUIRED.

>   string error_code = 1;

Reviewer 2: nit: s/error_code/code/

"error" is already in the name of the message type, so this seems to stutter as-is

Author: done

  // The error message associated with the above error code.
  // This field is OPTIONAL.

>   string message = 2;

Reviewer 2: I think the OPTIONAL/REQUIRED here are backwards. Messages should be REQUIRED, and error codes should be OPTIONAL. It's all opaque and requires human eyes/brains to parse - a message seems like the minimum bar to clear here. Otherwise if admin Dan reads one of these serialized VolumeHealthError objects and sees just a code (w/o a message) that's way less informative than a meaningful error message (w/o an error code).

Author: Sure

message NodeGetVolumeStatsResponse {
  // This field is OPTIONAL.
  repeated VolumeUsage usage = 1;

>   // This field is OPTIONAL.

Reviewer: ```suggestion
  // Information about the current condition of the volume.
  // This field is OPTIONAL.
  // This field MUST be specified if the VOLUME_CONDITION node capability is supported.
```

  // Units by which values are measured. This field is REQUIRED.
  Unit unit = 4;
}

> // VolumeCondition represents the current condition of a volume

Reviewer 2: nit: end sentence w/ a period

Author: Done

  // Units by which values are measured. This field is REQUIRED.
  Unit unit = 4;
}

// VolumeCondition represents the current health condition of a volume

> message VolumeCondition {

Reviewer: nit: What was wrong with `VolumeHealth` again? `VolumeCondition` seems weird to me.

Author: I think it is because we were using Conditions (condition_true, condition_false, condition_unknown) here in the previous version of the PR.  VolumeHealth condition sounds weird.  So James proposed to rename to VolumeCondition.

Now that we use .google.protobuf.BoolValue.  Maybe we should rename back to VolumeHealth?

@jdef what do you think?

Author: Changed from .google.protobuf.BoolValue to bool as "abnormal" is required now.

Reviewer 2: s/health condition/condition/ everywhere here and I'm happy with it. "health condition" is too awkward. Just get rid of the word "health" everywhere please.

What does healthy mean? I think it means different things to different people (as our many meetings seem to suggest). "Volume condition" - people actually have to read the docs to understand what this means. I like that. And it forces us to write decent documentation for "condition" fields.

"Indicates that the condition of a volume is not normal." -- this could use some polishing, it's too sparse. We don't need a novel here, but we need *something* more.  Maybe something along the lines of:

> *Normal* volumes are online (i.e. available for use) and operating optimally. An *abnormal* volume does not meet these criteria.

Author: "online" means volume is attached, but we also look at volumes that are provisioned but not attached yet.  So I changed to the following:
  
  // Normal volumes are available for use and operating optimally.
  // An abnormal volume does not meet these criteria.

  option (alpha_message) = true;

  // Indicates the health condition of a volume is not normal.

>   // This field is OPTIONAL. If not present, it indicates the

Reviewer: Should this field be REQUIRED since it is the first one?

Author: Makes sense.  If it is required, we can just use a bool?

  // Units by which values are measured. This field is REQUIRED.
  Unit unit = 4;
}

// VolumeCondition represents the current health condition of a volume
message VolumeCondition {
  option (alpha_message) = true;

>   // Indicates the health condition of a volume is not normal.

Author: Done

      GET_VOLUME_STATS = 2;
      // See VolumeExpansion for details.
      EXPAND_VOLUME = 3;

>       // Indicates the SP supports node level volume condition.

Reviewer 2: ditto re: "node level" - can we find better terminology than "controller/node-level"? maybe it's confusing for me because CSI plugin arch is flexible and so "level" doesn't really make sense for e.g. monolith plugins that combine both gRPC services.

suggestion: "Indicates that the Node service can report volume conditions."

Author: Done

      GET_VOLUME_STATS = 2;
      // See VolumeExpansion for details.
      EXPAND_VOLUME = 3;

>       // If Plugin implements VOLUME_CONDITION capability
>       // then it MUST implement NodeGetVolumeStats RPC
>       // call for fetching volume condition.

Reviewer: Same as controller capability, capability description should be a generic description (that way in the future if we add additional methods or fields that use this capability, this comment does not need to be updated).

```suggestion
      // Indicates the SP supports node level volume condition.
      // An SP MAY implement `VolumeCondition` in only the Node
      // Plugin, only the Controller Plugin, or both.
      // If `VolumeCondition` is implemented in both the Node and
      // Controller Plugins, it SHALL report from different perspectives.
      // If for some reason Node and Controller Plugins report
      // misaligned volume conditions, CO SHALL assume the worst case
      // is the truth.
      // Note that, for alpha, `VolumeCondition` is intended to be
      // informative for humans only, not for automation.
```

Author: Done

      GET_VOLUME_STATS = 2;
      // See VolumeExpansion for details.
      EXPAND_VOLUME = 3;

>       // If Plugin implements VOLUME_HEALTH capability

Reviewer 2: ditto: should probably appear in AlphaFeatures instead

Author: Done
